lib/apiservers/service/restapi/handlers/vch_list_get.go

```diff
@@ -0,0 +1,209 @@
+	validator, err := validateTarget(ctx, d)
+	if err != nil {
+		return nil, util.WrapError(400, err)
+	}
+
+	executor := management.NewDispatcher(validator.Context, validator.Session, nil, false)
+	vchs, err := executor.SearchVCHs(validator.ClusterPath)
+	if err != nil {
+		return nil, util.NewError(500, fmt.Sprintf("Failed to search VCHs in %s: %s", validator.ResourcePoolPath, err))
+	}
+
+	return vchsToModels(ctx, vchs, executor), nil
+}
+
+func buildData(url url.URL, user string, pass string, thumbprint *string, datacenter *string, computeResource *string) *data.Data {
+	d := data.Data{
+		Target: &common.Target{
+			URL:      &url,
+			User:     user,
+			Password: &pass,
+		},
+	}
+
+	if datacenter != nil {
+		// TODO: Convert ID to Name or update underlying code to accept ID
+		d.Target.URL.Path = *datacenter
+	}
+
+	if thumbprint != nil {
+		d.Thumbprint = *thumbprint
+	}
+
+	if computeResource != nil {
+		d.ComputeResourcePath = *computeResource
+	}
+
+	return &d
+}
+
+func validateTarget(ctx context.Context, d *data.Data) (*validate.Validator, error) {
+	if err := d.HasCredentials(); err != nil {
+		return nil, fmt.Errorf("Invalid Credentials: %s", err)
+	}
+
+	validator, err := validate.NewValidator(ctx, d)
+	if err != nil {
+		return nil, fmt.Errorf("Validation Error: %s", err)
+	}
+	// If dc is not set, and multiple datacenter is available, vic-machine ls will list VCHs under all datacenters.
+	validator.AllowEmptyDC()
+
+	_, err = validator.ValidateTarget(ctx, d)
+	if err != nil {
+		return nil, fmt.Errorf("Target validation failed: %s", err)
+	}
+	_, err = validator.ValidateCompute(ctx, d, false)
+	if err != nil {
+		return nil, fmt.Errorf("Compute resource validation failed: %s", err)
+	}
+
+	return validator, nil
+}
+
+// Copied from list.go, and appears to be present other places. TODO: deduplicate
+func upgradeStatusMessage(ctx context.Context, vch *vm.VirtualMachine, installerVer *version.Build, vchVer *version.Build) string {
+	if sameVer := installerVer.Equal(vchVer); sameVer {
+		return "Up to date"
+	}
+
+	upgrading, err := vch.VCHUpdateStatus(ctx)
+	if err != nil {
+		return fmt.Sprintf("Unknown: %s", err)
+	}
+	if upgrading {
+		return "Upgrade in progress"
+	}
+
+	canUpgrade, err := installerVer.IsNewer(vchVer)
+	if err != nil {
+		return fmt.Sprintf("Unknown: %s", err)
+	}
+	if canUpgrade {
+		return fmt.Sprintf("Upgradeable to %s", installerVer.ShortVersion())
+	}
+
+	oldInstaller, err := installerVer.IsOlder(vchVer)
+	if err != nil {
+		return fmt.Sprintf("Unknown: %s", err)
+	}
+	if oldInstaller {
+		return fmt.Sprintf("VCH has newer version")
+	}
+
+	// can't get here
+	return "Invalid upgrade status"
+}
+
+func vchsToModels(ctx context.Context, vchs []*vm.VirtualMachine, executor *management.Dispatcher) []*models.VCHListItem {
+	installerVer := version.GetBuild()
+	payload := make([]*models.VCHListItem, 0)
+	for _, vch := range vchs {
+		var version *version.Build
+		var dockerHost string
+		var adminPortal string
```

Comment:
you may want to declare them outside of the loop and reusing them if vchs potentially a big slice of vms

Reply:
If I declare them outside the loop, I need to re-zero them on each iteration.

Reply:
Just for curiosity I would take a look at the generated code, I suspect all these variable are allocated on stack.

Reply:
I'm not sure why it matters whether or not these three variables are allocated on the stack. Could you elaborate?